From 4262c9fb634094d021b7474a8794e3f26e7cf636 Mon Sep 17 00:00:00 2001
From: Simon Xue <xxm@rock-chips.com>
Date: Wed, 10 Jul 2019 12:02:39 +0800
Subject: [PATCH] iommu: hack for disable rockchip iommu

Get iommu_ops from platform bus of dev if domain is NULL to avoid
kernel crash

Change-Id: I5554b06a450805fb3812f1ade60f335efa4d2a4e
Signed-off-by: Simon Xue <xxm@rock-chips.com>
---
 drivers/iommu/iommu.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/iommu/iommu.c b/drivers/iommu/iommu.c
index 8c15c5980299..5c7954923ec8 100644
--- a/drivers/iommu/iommu.c
+++ b/drivers/iommu/iommu.c
@@ -1321,6 +1321,13 @@ static int __iommu_attach_device(struct iommu_domain *domain,
 				 struct device *dev)
 {
 	int ret;
+
+	/* Hack for disable iommu */
+	if (!domain) {
+		ret = dev->bus->iommu_ops->attach_dev(domain, dev);
+		return ret;
+	}
+
 	if ((domain->ops->is_attach_deferred != NULL) &&
 	    domain->ops->is_attach_deferred(domain, dev))
 		return 0;
-- 
2.35.3

